src/System.Runtime.Serialization.Xml/tests/XmlDictionaryWriterTest.cs

```diff
@@ -0,0 +1,104 @@
+        string actual;
+
+        int byteSize = 8;
+        byte[] bytes = new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 };
+
+        using (var ms = new AsyncMemoryStream())
+        {
+            var writer = XmlDictionaryWriter.CreateTextWriter(ms);
+            writer.WriteStartDocument();
+            writer.WriteStartElement("data");
+            var task = writer.WriteBase64Async(bytes, 0, byteSize);
+            task.Wait();
+            writer.WriteEndElement();
+            writer.WriteEndDocument();
+            writer.Flush();
+            ms.Position = 0;
+            var sr = new StreamReader(ms);
+            actual = sr.ReadToEnd();
+        }
+
+        Assert.StrictEqual(expect, actual);
+    }
+
+    [Fact]
+    public static void XmlBaseWriter_FlushAsync()
+    {
+        string expect = "<?xml version=\"1.0\" encoding=\"utf-8\"?><data>AQIDBAUGBwg=</data>";
+        string actual;
+
+        int byteSize = 8;
+        byte[] bytes = new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 };
+
+        using (var ms = new AsyncMemoryStream())
+        {
+            var writer = XmlDictionaryWriter.CreateTextWriter(ms);
+            writer.WriteStartDocument();
+            writer.WriteStartElement("data");
+            writer.WriteBase64(bytes, 0, byteSize);
+            writer.WriteEndElement();
+            writer.WriteEndDocument();
+            var task = writer.FlushAsync();
+            task.Wait();
+            ms.Position = 0;
+            var sr = new StreamReader(ms);
+            actual = sr.ReadToEnd();
+        }
+
+        Assert.StrictEqual(expect, actual);
+    }
+
+    [Fact]
+    public static void XmlBaseWriter_WriteStartEndElementAsync()
+    {
+        string expect = "<?xml version=\"1.0\" encoding=\"utf-8\"?><data>AQIDBAUGBwg=</data>";
+        string actual;
+
+        int byteSize = 8;
+        byte[] bytes = new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 };
+
+        using (var ms = new AsyncMemoryStream())
+        {
+            var writer = XmlDictionaryWriter.CreateTextWriter(ms);
+            writer.WriteStartDocument();
+            // NOTE: the async method has only one overload that takes 3 params
+            var t1 = writer.WriteStartElementAsync(null, "data", null);
+            t1.Wait();
+            writer.WriteBase64(bytes, 0, byteSize);
+            var t2 = writer.WriteEndElementAsync();
+            t2.Wait();
+            writer.WriteEndDocument();
+            writer.Flush();
+            ms.Position = 0;
+            var sr = new StreamReader(ms);
+            actual = sr.ReadToEnd();
+        }
+
+        Assert.StrictEqual(expect, actual);
+    }
+
+    public class AsyncMemoryStream : MemoryStream
+    {
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            await Task.Delay(10).ConfigureAwait(false);
+            await base.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+    }
+}
```

Comment:
It'd be good to get an additional test to (on a basic level) ensure that the XmlDictionaryAsyncCheckWriter is indeed throwing when async behaviours happen at the same time.